test/unit/src/ShardFunctions.test.ts

```diff
@@ -0,0 +1,2777 @@
+
+      const result = ShardFunctions.partitionToAddressRange2(largeShardGlobals, partition)
+
+      expect(result).toBeDefined()
+      expect(result.partition).toBe(partition)
+      expect(result.endAddr).toBe(4294967295) // This is 0xFFFFFFFF
+    })
+
+    it('should adjust endAddr to prevent rounding errors at boundary', () => {
+      // Create a shard global with a number of partitions that could trigger rounding errors
+      const roundingShardGlobals = { ...mockShardGlobals, numPartitions: 7 }
+      const partition = roundingShardGlobals.numPartitions - 1
+
+      const result = ShardFunctions.partitionToAddressRange2(roundingShardGlobals, partition)
+
+      expect(result).toBeDefined()
+      // The last partition should reach all the way to the maximum value
+      expect(result.endAddr).toBe(4294967295) // 0xFFFFFFFF
+    })
+
+    it('should set the correct string format for low and high addresses', () => {
+      const partition = 1
+
+      const result = ShardFunctions.partitionToAddressRange2(mockShardGlobals, partition)
+
+      expect(result.low.length).toBe(64)
+      expect(result.high.length).toBe(64)
+      // Check that the first 8 characters are the hex representation of startAddr and endAddr
+      expect(result.low.substring(0, 8)).toBe(('00000000' + result.startAddr.toString(16)).slice(-8))
+      expect(result.high.substring(0, 8)).toBe(('00000000' + result.endAddr.toString(16)).slice(-8))
+      // Check the padding
+      expect(result.low.substring(8)).toBe('0'.repeat(56))
+      expect(result.high.substring(8)).toBe('f'.repeat(56))
+    })
+
+    it('should handle endPartition equal to numPartitions', () => {
+      // Setup with partitionMax to force endPartition to equal numPartitions
+      const partition = 0
+      const partitionMax = mockShardGlobals.numPartitions - 1
+
+      const result = ShardFunctions.partitionToAddressRange2(mockShardGlobals, partition, partitionMax)
+
+      expect(result).toBeDefined()
+      expect(result.partitionEnd).toBe(partitionMax)
+    })
+  })
+
+  describe('#addressNumberToPartition', () => {
+    it('should convert address number to partition', () => {
+      const addressNum = 0x12345678
+
+      const result = ShardFunctions.addressNumberToPartition(mockShardGlobals, addressNum)
+
+      expect(typeof result).toBe('number')
+    })
+
+    it('should handle max address number', () => {
+      const addressNum = 0xffffffff
+
+      const result = ShardFunctions.addressNumberToPartition(mockShardGlobals, addressNum)
+
+      expect(result).toBe(mockShardGlobals.numPartitions - 1)
+    })
+  })
+
+  describe('#addressToPartition', () => {
+    it('should convert address to partition', () => {
+      const address = '12345678deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef'
+
+      const result = ShardFunctions.addressToPartition(mockShardGlobals, address)
+
+      expect(result).toBeDefined()
+      expect(typeof result.homePartition).toBe('number')
```

Comment:
can we also validate the value in addition to type? is the output deterministic ?